--- originals-4502/Interpreter.c	2007-01-10 15:57:51.000000000 -0500
+++ ../sys/Interpreter.c	2007-01-10 15:58:24.000000000 -0500
@@ -40,6 +40,7 @@
 #include "praatP.h"
 #include "praat_script.h"
 #include "Formula.h"
+#include "../jscript/mono_embed.h"
 
 #define Interpreter_WORD 1
 #define Interpreter_REAL 2
@@ -595,6 +596,15 @@
 }
 
 int Interpreter_run (Interpreter me, char *text) {
+	if (strncmp(text, "//javascript", 12) == 0) {
+		mono_embed_run_praat_script(me, text);
+		return 1;
+	}
+	if (strncmp(text, "#python", 7) == 0) {
+		mono_embed_run_praat_script(me, text);
+		return 1;
+	}
+
 	static char valueString [30000];   /* To divert the info. */
 	char *command = text, command2 [28000];
 	char **lines = NULL;
