diff -ur -x '*.[oa]' sources_5308/makefile sources_current/makefile
--- sources_5308/makefile	2012-01-14 07:28:16.000000000 -0500
+++ sources_current/makefile	2012-03-07 13:10:54.000000000 -0500
@@ -31,6 +31,7 @@
 	$(MAKE) -C artsynth
 	$(MAKE) -C contrib/ola
 	$(MAKE) -C main main_Praat.o $(ICON)
+	$(MAKE) -C scripting
 	$(LINK) -o $(EXECUTABLE) main/main_Praat.o $(MAIN_ICON) fon/libfon.a \
 		contrib/ola/libOla.a artsynth/libartsynth.a \
 		FFNet/libFFNet.a gram/libgram.a EEG/libEEG.a \
@@ -40,6 +41,7 @@
 		external/espeak/libespeak.a external/portaudio/libportaudio.a \
 		external/flac/libflac.a external/mp3/libmp3.a \
 		external/glpk/libglpk.a external/gsl/libgsl.a \
+		scripting/scripting.o scripting/python.o scripting/util.o `python-config --ldflags` \
 		$(LIBS)
 
 clean:
@@ -62,5 +64,6 @@
 	$(MAKE) -C FFNet clean
 	$(MAKE) -C artsynth clean
 	$(MAKE) -C contrib/ola clean
+	$(MAKE) -C scripting clean
 	# $(MAKE) -C main clean
 	$(RM) praat
diff -ur -x '*.[oa]' sources_5308/makefiles/makefile.defs.linux sources_current/makefiles/makefile.defs.linux
--- sources_5308/makefiles/makefile.defs.linux	2011-09-08 07:55:55.000000000 -0400
+++ sources_current/makefiles/makefile.defs.linux	2012-03-07 13:10:06.000000000 -0500
@@ -13,7 +13,7 @@
 
 LINK = g++
 
-EXECUTABLE = praat
+EXECUTABLE = praat-py
 
 LIBS = `pkg-config --libs gtk+-2.0` -lm -lasound
 
diff -ur -x '*.[oa]' sources_5308/makefiles/makefile.defs.mingw32 sources_current/makefiles/makefile.defs.mingw32
--- sources_5308/makefiles/makefile.defs.mingw32	2011-09-14 11:15:37.000000000 -0400
+++ sources_current/makefiles/makefile.defs.mingw32	2012-03-07 13:11:56.000000000 -0500
@@ -15,7 +15,7 @@
 
 LINK = /mingw32/bin/g++
 
-EXECUTABLE = Praat.exe
+EXECUTABLE = Praat-Py.exe
 
 LIBS = -L/mingw32/lib -lwinmm -lwsock32 -lcomctl32 -lole32 -lgdi32 -lgdiplus -lcomdlg32 -static-libgcc -static-libstdc++ -mwindows
 #LIBS = -L/mingw32/lib -lwinmm -lwsock32 -lcomctl32 -lole32 -lgdi32 -lgdiplus -lcomdlg32 /mingw32/lib/QTMLClient.lib -static-libgcc -static-libstdc++ -mwindows
diff -ur -x '*.[oa]' sources_5308/sys/Interpreter.cpp sources_current/sys/Interpreter.cpp
--- sources_5308/sys/Interpreter.cpp	2012-03-03 15:54:10.000000000 -0500
+++ sources_current/sys/Interpreter.cpp	2012-03-07 13:34:03.000000000 -0500
@@ -61,6 +61,7 @@
 #include "Formula.h"
 #include "praat_version.h"
 #include "UnicodeData.h"
+#include "../scripting/scripting.h"
 
 #define Interpreter_WORD 1
 #define Interpreter_REAL 2
@@ -628,7 +629,14 @@
 	}
 }
 
+void Interpreter_setArgv (Interpreter me, const wchar **argv) {
+	my argv = (wchar **)argv; // loss of const can't be a good thing...
+}
+
 void Interpreter_run (Interpreter me, wchar *text) {
+	if (scripting_run_praat_script(me, text, my argv))
+		return;
+	
 	autoNUMvector <wchar *> lines;   // not autostringvector, because the elements are reference copies
 	long lineNumber = 0;
 	bool assertionFailed = false;
diff -ur -x '*.[oa]' sources_5308/sys/Interpreter.h sources_current/sys/Interpreter.h
--- sources_5308/sys/Interpreter.h	2011-08-15 09:42:29.000000000 -0400
+++ sources_current/sys/Interpreter.h	2012-11-11 13:05:03.000000000 -0500
@@ -55,6 +55,7 @@
 		long labelLines [1+Interpreter_MAXNUM_LABELS];
 		wchar dialogTitle [1+100], procedureNames [1+Interpreter_MAX_CALL_DEPTH] [100];
 		SortedSetOfString variables;
+		wchar **argv;
 		bool running, stopped;
 	// overridden methods:
 	protected:
@@ -78,6 +79,7 @@
 void Interpreter_stringExpression (Interpreter me, const wchar *expression, wchar **value);
 void Interpreter_numericArrayExpression (Interpreter me, const wchar *expression, struct Formula_NumericArray *value);
 void Interpreter_anyExpression (Interpreter me, const wchar *expression, struct Formula_Result *result);
+void Interpreter_setArgv (Interpreter me, const wchar **argv);
 
 InterpreterVariable Interpreter_hasVariable (Interpreter me, const wchar *key);
 InterpreterVariable Interpreter_lookUpVariable (Interpreter me, const wchar *key);
diff -ur -x '*.[oa]' sources_5308/sys/praat.cpp sources_current/sys/praat.cpp
--- sources_5308/sys/praat.cpp	2012-02-18 07:04:25.000000000 -0500
+++ sources_current/sys/praat.cpp	2012-03-07 13:17:44.000000000 -0500
@@ -1193,13 +1193,16 @@
 	if (Melder_batch) {
 		#if defined (UNIX) || defined (macintosh) || defined (_WIN32) && defined (CONSOLE_APPLICATION)
 			MelderString_empty (& theCurrentPraatApplication -> batchName);
+			theCurrentPraatApplication -> argv = (wchar**)malloc (sizeof (wchar*) * argc);
 			for (unsigned int i = iarg_batchName; i < argc; i ++) {
 				int needsQuoting = strchr (argv [i], ' ') != NULL && (i == iarg_batchName || i < argc - 1);
 				if (i > 1) MelderString_append (& theCurrentPraatApplication -> batchName, L" ");
 				if (needsQuoting) MelderString_append (& theCurrentPraatApplication -> batchName, L"\"");
 				MelderString_append (& theCurrentPraatApplication -> batchName, Melder_peekUtf8ToWcs (argv [i]));
 				if (needsQuoting) MelderString_append (& theCurrentPraatApplication -> batchName, L"\"");
+				theCurrentPraatApplication -> argv[i-iarg_batchName] = Melder_peekUtf8ToWcs (argv [i]);
 			}
+			theCurrentPraatApplication -> argv[argc-iarg_batchName] = NULL;
 		#elif defined (_WIN32)
 			MelderString_copy (& theCurrentPraatApplication -> batchName, Melder_peekUtf8ToWcs (argv [3]));
 		#endif
@@ -1417,7 +1420,7 @@
 			}
 		} else {
 			try {
-				praat_executeScriptFromFileNameWithArguments (theCurrentPraatApplication -> batchName.string);
+				praat_executeScriptFromFileNameWithArguments2 (theCurrentPraatApplication -> batchName.string, (const wchar**)theCurrentPraatApplication -> argv);
 				praat_exit (0);
 			} catch (MelderError) {
 				/*
diff -ur -x '*.[oa]' sources_5308/sys/praat.h sources_current/sys/praat.h
--- sources_5308/sys/praat.h	2011-09-13 11:43:34.000000000 -0400
+++ sources_current/sys/praat.h	2012-03-07 13:27:08.000000000 -0500
@@ -144,6 +144,7 @@
 	int batch;   /* Was the program called from the command line? */
 	GuiObject topShell;   /* The application shell: parent of editors and standard dialogs. */
 	ManPages manPages;
+	wchar **argv; /* NULL-terminated list of command-line arguments including the file name. */
 } structPraatApplication, *PraatApplication;
 typedef struct {   /* Readonly */
 	int n;	 /* The current number of objects in the list. */
diff -ur -x '*.[oa]' sources_5308/sys/praat_script.cpp sources_current/sys/praat_script.cpp
--- sources_5308/sys/praat_script.cpp	2011-09-11 17:11:54.000000000 -0400
+++ sources_current/sys/praat_script.cpp	2012-03-07 13:31:28.000000000 -0500
@@ -412,6 +412,10 @@
 }
 
 void praat_executeScriptFromFile (MelderFile file, const wchar *arguments) {
+	praat_executeScriptFromFile2 (file, arguments, NULL);
+}
+
+void praat_executeScriptFromFile2 (MelderFile file, const wchar_t *arguments, const wchar **argv) {
 	try {
 		autostring text = MelderFile_readText (file);
 		autoMelderFileSetDefaultDir dir (file);   // so that relative file names can be used inside the script
@@ -421,6 +425,9 @@
 			Interpreter_readParameters (interpreter.peek(), text.peek()); therror
 			Interpreter_getArgumentsFromString (interpreter.peek(), arguments); therror
 		}
+		if (argv) {
+			Interpreter_setArgv (interpreter.peek(), argv);
+		}
 		Interpreter_run (interpreter.peek(), text.peek()); therror
 	} catch (MelderError) {
 		Melder_throw ("Script ", file, " not completed.");
@@ -428,6 +435,10 @@
 }
 
 void praat_executeScriptFromFileNameWithArguments (const wchar *nameAndArguments) {
+	praat_executeScriptFromFileNameWithArguments2 (nameAndArguments, NULL);
+}
+
+void praat_executeScriptFromFileNameWithArguments2 (const wchar_t *nameAndArguments, const wchar **argv) {
 	wchar_t path [256];
 	const wchar_t *p, *arguments;
 	structMelderFile file = { 0 };
@@ -452,7 +463,7 @@
 		if (*arguments == ' ') arguments ++;
 	}
 	Melder_relativePathToFile (path, & file);
-	praat_executeScriptFromFile (& file, arguments);
+	praat_executeScriptFromFile2 (& file, arguments, argv);
 }
 
 void praat_executeScriptFromText (wchar *text) {
diff -ur -x '*.[oa]' sources_5308/sys/praat_script.h sources_current/sys/praat_script.h
--- sources_5308/sys/praat_script.h	2011-08-15 20:55:24.000000000 -0400
+++ sources_current/sys/praat_script.h	2012-03-07 13:29:08.000000000 -0500
@@ -24,7 +24,9 @@
 void praat_executeCommand (Interpreter me, wchar *command);
 void praat_executeCommandFromStandardInput (const char *programName);
 void praat_executeScriptFromFile (MelderFile file, const wchar *arguments);
+void praat_executeScriptFromFile2 (MelderFile file, const wchar_t *arguments, const wchar **argv);
 void praat_executeScriptFromFileNameWithArguments (const wchar *nameAndArguments);
+void praat_executeScriptFromFileNameWithArguments2 (const wchar_t *nameAndArguments, const wchar **argv);
 void praat_executeScriptFromText (wchar *text);
 void praat_executeScriptFromDialog (Any dia);
 void DO_praat_runScript (UiForm sendingForm, const wchar *sendingString, Interpreter interpreter_dummy, const wchar *invokingButtonTitle, bool modified, void *dummy);
